### PR TITLE
Memory Management Milestone

### PR DESCRIPTION
**Pull request by student-of**
Memory Management Team Milestone
Includes:
* `MemoryManager` which acts as the central API for static/GC memory allocation
* Various heap allocator implementations which may be used in GC/non-GC contexts
* Mark-and-sweep GC  (`MnS`)  generic over a heap implementation
* Code generation integration for variable/temporary rooting and GC memory allocation
* Proposal updates

Currently, the GC and static allocator use separate bump allocators (e.g. no freeing) to allocate memory so we have feature parity with the base-line WASM compiler. All current tests pass.

### CODE REVIEW

**Review by jpolitz — COMMENTED** [review]
I'm struggling a bit with this. I agree with a lot of the identified challenges – lack of access to WASM stack means that in the current compiler root tracking is quite difficult.

I think that if we require that we wrap everything in `codeGenTemp...` stuff, folks will forget and things will be brittle.

In addition, this isn't a GC issue just a code issue, the leading list of `$$` variables in a function is problematic, and I think in cases of nested expressions, broken.

We need some kind of API/structure for temporaries other than just concatenating more and more locals, or concatenating extra statements around locals. This API can hook into the memory manager.

I think the full consequences of this are pretty challenging:

- Without deeper analysis, _every_ non-bool temporary needs storage that isn't just stack storage. This would be a code-style change throughout codegen, instead of relying on implicit stacking, add helpers to track a set of generated temporaries and only use those, then splat out the relevant temporaries at the top of each function definition after compiling the body.
- In theory, the above can be done programmatically and statically on the generated WASM. Due to WASM's restrictions, each function body has a determinable stack depth, and we could simply name each stack location with a numbered variable name, and track those on updates.
- If we had a flatter IR, we could do the temporary marking and storage only around allocating expressions (strings, lists, construct, function calls, and so on). Not sure we can move to the flatter IR in a week :-/

I pushed my merge attempt to https://github.com/ucsd-cse231-w21/chocopy-wasm-compiler/tree/merging-mm, which is failing a bunch of tests. Maybe there's a simple tweak to get this working, but I'm not sure.

To do this in a week, I honestly think the best bet would be to update allocations to use the gc-defined alloc* functions, then try and post-process the generated WASM in each function body to find and save all the locals on updates. That will avoid synchronizing with all code updates to codegen and firming that API up at the same time as implementing more GC _and_ features.

File: compiler.ts [hunk 1]
```diff
@@ -122,23 +127,58 @@ function codeGenStmt(stmt: Stmt<Type>, env: GlobalEnv): Array<string> {
...
       return valStmts;
     case "assignment":
       throw new Error("Destructured assignment not implemented");
     case "assign":
       var valStmts = codeGenExpr(stmt.value, env);
       if (env.locals.has(stmt.name)) {
-        return valStmts.concat([`(local.set $${stmt.name})`]);
+        // NOTE(alex:mm): removeLocal/addLocal calls are necessary b/c
+        //   MemoryManager cannot scan the WASM stack directly and
+        //   must maintain a list of local variable pointers
```
[the remark this answers is not in the document]
Comment (student-of) [reply compiler.ts:158]: Yeah, I have a planned fix for that in a local branch (essentially a stack of the last indice of the caller temporary frame)

[the remark this answers is not in the document]
Comment (student-of) [reply compiler.ts:158]: Ah, you mean the WASM stack frame, not metadata. IIRC, a bunch of languages with and without GC support already do this (like C).
On the memory management side, we can actually easily do that if you wish to support it.

I haven't done any bench marking but I suspect it shouldn't be too bad. It's only executing a function call per variable update (which can be optimized to possible-pointer values only), and WASM was designed for fast JS interop, no?

**Comment by student-of** [comment]
> I think that if we require that we wrap everything in codeGenTemp... stuff, folks will forget and things will be brittle.
`codeGenTempGuard()` only *needs* to be called at the statement level, so in that regard, I don't think it is that brittle. If people want to do it at the expression level, that can be considered an optimization. Temporaries usually only live as long as the statement and `codeGenTempGuard()` enforces that (barring some bugs I need to fix).

>  the leading list of $$ variables in a function is problematic, and I think in cases of nested expressions, broken
Yeah, I believe so too. I think we need some sort of analysis pass for creating temporary local variables.

> we could simply name each stack location with a numbered variable name,
I actually need to fix `addLocal(value)/removeLocal(value)` to be `addLocal(index, value)/removeLocal(index)` (the current version is incorrect when updating aliasing local variables). With that new type signature, we can _emulate_ tracking WASM locals purely via stack location.

> If we had a flatter IR
The flatter IR would be ideal but also runs into the `brittle` issue without a secondary pass to insert the guards.

> post-process the generated WASM in each function body to find and save all the locals
The easiest way to do this would be to scan `local.set` instructions for non-`$$` names. Care needs to be taken such that this post-process step doesn't occur after optimization (hypothetically, a local variable may be only needed as a temporary stack object instead of a stack slot). However, statements still need to be wrapped with `codeGenTempGuard()` unless we use AST data to guide the post-processing.

**Comment by student-of** [comment]
> To do this in a week

Would you like us to perform the changes for the milestone merge? Or do it post-merge?

File: compiler.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jpolitz) [reply compiler.ts:158]: To elaborate – it seems painful to call a function on every variable assignment. I haven't measured, but that seems really unlikely to be cheap.

So another alternative is to gather the needed locals and temporaries by looking at the body of the function (at the source level, but maybe also figuring out needed temps), and then *heap-allocate* that structure. Then at the entry and exit of each function there can be a push/pop of a reference to that structure, unconditionally. Variable updates become heap stores, and variable lookups become loads. That part is more expensive, certainly.

So maybe somehow constants (unassigned variables, or assigned-once variables) can get a different treatment. But calling a function when there's a variable update in a loop just seems rough.

**Comment by jpolitz** [comment]
I can't get the milestone merge to work. See https://github.com/ucsd-cse231-w21/chocopy-wasm-compiler/tree/merging-mm.

In terms of the course project, this is all good stuff/conversation.

In terms of actually merging working code along with what everyone else is doing, I'm looking for reasonable hacks. You might have to be tracking some of the heap allocation happening on other PRs to have a sense of what to do with some of this.

**Comment by student-of** [comment]
I'm working here: https://github.com/SasyaReddi/chocopy-wasm-compiler/tree/merging-mm

Around 13 tests are failing now.

I've rewritten most of the allocation codegen but having a bit of trouble with lists and bigints. Maybe half of the failing tests are caused by primitive/size confusion (I see the correct byte embedded in the wrong 32-bit number for instance). One or two tests are failing due to bad temporary/stack rooting codegen.

I will not be splitting the rooting into a separate pass until I have a working version to test against.

**Comment by student-of** [comment]
Alright, only one test failing now involving list allocations. When that's resolved, want me to merge into `memory-management` or open a new pull request?

**Comment by jpolitz** [comment]
Just merge into this branch and we can do the PR from here w/all the history. Thanks!

**Comment by student-of** [comment]
Quick side note: it looks like the usages of generated `$` variables *that I've seen* are correct in the presence of nested expressions (including a fix for class construction).

The common pattern is that the variables are just used to dump the value to the stack X amount of times before generating code for inner expressions and never used afterwards.

**Comment by jpolitz** [comment]
What about something like this:

https://github.com/ucsd-cse231-w21/chocopy-wasm-compiler/pull/65/files/7be58cbe9a4ff568b2c7a88c69e65f30cb95abd3#diff-75bfbd17ff828cc0800075e681515a281a5b67d83348a11c83df27c4f2e3f3b4R407

**Comment by jpolitz** [comment]
My question there is an honest question, but I'm concerned that the nested expression could overwrite `$$string_address`, causing a wrong lookup.

**Comment by student-of** [comment]
Alright, merged the fixes including swapping all heap allocations to use `gcalloc`. All tests pass.

Not included: a second pass that inserts the WASM instructions.

**Comment by student-of** [comment]
Think this is ready to merge?

**Comment by jpolitz** [comment]
OK, so I missed an argument on `withDefines`, but there's also something going on with the `range` helper that is probably messing with memory. I get `range(5)` and `range(10)` tests failing with `npm test`, but if I run them individually I get successes, and if I just run `npm test -- -g "range"` the first succeeds and the second fails. So, uh, I think something in a memory shared across runs is messed up.

**Comment by student-of** [comment]
Ah, that's my bad. Looks like `$range` uses old heap allocation instructions. As a side note, it would be nice if it was moved into a separate file or something...

**Comment by student-of** [comment]
Uncovered more issues I missed. There's code in `tests/import-object.test.ts` that sets the first 4 bytes of linear memory. Removing it revealed more failing tests and allocation code I missed.

**Comment by student-of** [comment]
Currently blocked on some dictionary issues causing an infinite loop somewhere on bracket lookups.

**Comment by student-of** [comment]
@jpolitz Think this is ready now.

**Pull request merged**